admin/aws.py

```diff
@@ -0,0 +1,203 @@
+    :ivar bytes prefix: Prefix to add to each key to delete.
+    :ivar list keys: List of keys to be deleted.
+    """
+
+
+@sync_performer
+def perform_delete_s3_keys(dispatcher, intent):
+    s3 = boto.connect_s3()
+    bucket = s3.get_bucket(intent.bucket)
+    bucket.delete_keys(
+        [intent.prefix + key
+         for key in intent.keys])
+
+
+@attributes([
+    "source_bucket",
+    Attribute("source_prefix", default_value=""),
+    "destination_bucket",
+    Attribute("destination_prefix", default_value=""),
+    "keys",
+])
+class CopyS3Keys(object):
+    """
+    Copy a list of keys from one S3 bucket to another.
+
+    :ivar bytes source_bucket: Name of bucket to copy keys from.
+    :ivar bytes source_prefix: Prefix to add to each key to in
+        ``source_bucket``.
+    :ivar bytes destination_bucket: Name of bucket to copy keys to.
+    :ivar bytes destination_prefix: Prefix to add to each key to in
+        ``destination_bucket``.
+    :ivar list keys: List of keys to be copied.
+    """
+
+
+@sync_performer
+def perform_copy_s3_keys(dispatcher, intent):
+    s3 = boto.connect_s3()
+    destination_bucket = s3.get_bucket(intent.destination_bucket)
+    for key in intent.keys:
+        destination_bucket.copy_key(
+            new_key_name=intent.destination_prefix + key,
+            src_bucket_name=intent.source_bucket,
+            src_key_name=intent.source_prefix + key)
+
+
+@attributes([
+    "bucket",
+    "prefix",
+])
+class ListS3Keys(object):
+    """
+    List the S3 keys in a bucket.
+
+    Note that returns a list with the prefixes stripped.
+
+    :ivar bytes bucket: Name of bucket to list keys from.
+    :ivar bytes prefix: Prefix of keys to be listed.
+    """
+
+
+@sync_performer
+def perform_list_s3_keys(dispatcher, intent):
+    s3 = boto.connect_s3()
+    bucket = s3.get_bucket(intent.bucket)
+    return {key.name[len(intent.prefix):]
+            for key in bucket.list(intent.prefix)}
+
+
+boto_dispatcher = TypeDispatcher({
+    UpdateS3RoutingRule: perform_update_s3_routing_rule,
+    ListS3Keys: perform_list_s3_keys,
+    DeleteS3Keys: perform_delete_s3_keys,
+    CopyS3Keys: perform_copy_s3_keys,
+    CreateCloudFrontInvalidation: perform_create_cloudfront_invalidation,
+})
+
+
+@attributes([
+    Attribute('routing_rules'),
+    Attribute('s3_buckets')
+])
+class FakeAWS(object):
+    def __init__(self):
+        self.cloudfront_invalidations = []
+
+    @sync_performer
+    def _perform_update_s3_routing_rule(self, dispatcher, intent):
+        old_target = self.routing_rules[intent.bucket][intent.prefix]
+        self.routing_rules[intent.bucket][intent.prefix] = intent.target_prefix
+        return old_target
+
+    @sync_performer
+    def _perform_create_cloudfront_invalidation(self, dispathcer, intent):
+        self.cloudfront_invalidations.append(intent)
+
+    @sync_performer
+    def _perform_delete_s3_keys(self, dispathcer, intent):
```

Comment:
This happens a few times and probably warrants a find/replace. I will stop commenting on it.